apps/settings/main.py

```diff
@@ -1,236 +1,158 @@
-                    pb.run_in_background()
-                pb.progress -= progress_per_step
-            sleep(1) # Needed here so that 1) the progressbar goes to 0 2) run_in_background launches the thread before the final stop() call
-            #TODO: add a way to pause the Refresher
-            pb.stop()
-            logger.info("Update failed")
-            PrettyPrinter("Update failed, try again later?", i, o, 3)
+                    bar.message = step.name
+                    step.do()
+                    finished_steps.append(step)
+                    bar.progress = len(finished_steps) / len(self.steps)
+                    bar.refresh()
+                    sleep(0.5)  # so the user has time to read what happens
+                except Exception as e:
+                    if DialogBox("yn", self.i, self.o,
+                                 message="'{}'failed\ncancel update?".format(step.name)).activate():
+                        self.rollback_update(bar, e, finished_steps, step)
+                        return -1
+                    else:
+                        pass
+
+    def rollback_update(self, bar, e, finished_steps, step):
+        logger.error("Error updating step '{}'".format(step.name))
+        logger.exception(e)
+        Printer([step.name, 'update failed'], self.i, self.o)
+        bar.message = "reverting"
+        for i, passed in enumerate(finished_steps):
+            passed.undo()
+            bar.progress = (len(finished_steps) - i) / len(self.steps)
+            bar.refresh()
+            sleep(0.5)
+
+    def create_steps(self, branch="master"):
+        tmp_dir = tempfile.mkdtemp(prefix='zpui')
+        os.chmod(tmp_dir, 0o777)
+        old_cwd = os.getcwd()
+        change_cwd = UpdateStep("opening tmp dir", lambda: os.chdir(tmp_dir), lambda: os.chdir(old_cwd))
+        self.steps.append(change_cwd)
+
+        git_copy = UpdateStep(
+            "copying repo",
+            "git clone {src_dir} {dir}/".format(src_dir=ZPUI_INSTALL_DIR, dir=tmp_dir, branch=branch)
+        )
+        self.steps.append(git_copy)
+
+        git_pull = UpdateStep(
+            "pulling git",
+            "git pull origin {branch} --ff-only".format(dir=tmp_dir, branch=branch)
+        )
+        self.steps.append(git_pull)
+
+        pip_install = UpdateStep("installing deps",
+                                 "pip2 install -r requirements.txt",
+                                 "pip2 install -r {src_dir}requirements.txt".format(src_dir=ZPUI_INSTALL_DIR)
+                                 )
+        self.steps.append(pip_install)
+
+        run_tests = UpdateStep(
+            "running tests",
+            "python2 -B -m pytest --doctest-modules -v --doctest-ignore-import-errors --ignore=output/drivers "
+            "--ignore=input/drivers --ignore=apps/hardware_apps/status/ --ignore=apps/example_apps/fire_detector "
+            "--ignore=apps/test_hardware",
+            accepted_return_code=0
+        )
```

Comment:
```
['python2', '-B', '-m', 'pytest', '--doctest-modules', '-v', '--doctest-ignore-import[17/1812]
 '--ignore=output/drivers', '--ignore=input/drivers', '--ignore=apps/hardware_apps/status/', '
--ignore=apps/example_apps/fire_detector', '--ignore=apps/test_hardware']
==================================== test session starts =====================================
platform linux2 -- Python 2.7.9, pytest-3.3.1, py-1.5.2, pluggy-0.6.0 -- /usr/bin/python2
cachedir: .cache
rootdir: /tmp/zpuiMi8zqD, inifile:
collected 13 items / 2 errors / 5 skipped

=========================================== ERRORS ===========================================
_________________________ ERROR collecting ui/tests/test_checkbox.py _________________________
ImportError while importing test module '/tmp/zpuiMi8zqD/ui/tests/test_checkbox.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ui/tests/test_checkbox.py:9: in <module>
    from checkbox import Checkbox
E   ImportError: No module named checkbox
______________________ ERROR collecting ui/tests/test_config_manager.py ______________________
ImportError while importing test module '/tmp/zpuiMi8zqD/ui/tests/test_config_manager.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
ui/tests/test_config_manager.py:9: in <module>
    from config_manager import UIConfigManager
E   ImportError: No module named config_manager
!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!
============================ 5 skipped, 2 error in 17.87 seconds =============================
```

Reply:
Merging latest changes in `devel` into your branch should fix that error. If not, before testing, remove all *.pyc files and the `__pycache__` folder from `ui/tests` .